common/filesystem.h

```diff
@@ -0,0 +1,1374 @@
+// removed recursively when destroyed.
+//
+// This can be used to ensure removal of a directory, and also exposes the
+// absolute path of the directory.
+//
+// As removal may encounter errors, unless the desired behavior is a
+// check-failure, users should explicitly move and call `Remove` at the end of
+// lifetime and handle any resultant errors.
+class RemovingDir : public Dir {
+ public:
+  RemovingDir() = default;
+  RemovingDir(RemovingDir&& arg) = default;
+  explicit RemovingDir(Dir d, std::filesystem::path abs_path)
+      : Dir(std::move(d)), abs_path_(std::move(abs_path)) {}
+  ~RemovingDir();
+  auto operator=(RemovingDir&& rhs) -> RemovingDir& = default;
+
+  auto abs_path() const -> const std::filesystem::path& { return abs_path_; }
+
+  // Releases the directory from being removed and returns just the underlying
+  // owning handle.
+  auto Release() && -> Dir { return std::move(*this); }
+
+  // Removes the directory immediately and surfaces any errors encountered.
+  auto Remove() && -> ErrorOr<Success, PathError>;
+
+ private:
+  friend Dir;
+
+  std::filesystem::path abs_path_;
+};
+
+// A named entry in a directory.
+//
+// This provides access to the scanned data when reading the entries of the
+// directory. It can only be produced by iterating over a `DirRef::Reader`.
+class DirRef::Entry {
+ public:
+  // The name of the entry.
+  //
+  // This is exposed as a null-terminated C-string as that is the most common
+  // representation.
+  auto name() const -> const char* { return dent_->d_name; }
+
+  // Test whether the entry is known to be a directory.
+  //
+  // An empty optional indicates it is unknown what kind of entry this is, while
+  // an engaged optional contains a bool indicating whether it is known a
+  // directory or known a non-directory entry.
+  auto is_dir() const -> std::optional<bool> {
+    return dent_->d_type == DT_UNKNOWN ? std::nullopt
+                                       : std::optional(dent_->d_type == DT_DIR);
+  }
+
+ private:
+  friend Dir::Reader;
+  friend Dir::Iterator;
+
+  Entry() = default;
+  explicit Entry(dirent* dent) : dent_(dent) {}
+
+  dirent* dent_ = nullptr;
+};
+
+// An iterator into a `DirRef::Reader`, used for walking the entries in a
+// directory.
+//
+// Most of the work of iterating a directory is done when constructing the
+// `Reader`, when constructing the beginning iterator, or when incrementing the
+// iterator.
+class DirRef::Iterator
+    : public llvm::iterator_facade_base<Iterator, std::input_iterator_tag,
+                                        const Entry> {
+ public:
+  // Default construct a general end iterator.
+  Iterator() = default;
+
+  auto operator==(const Iterator& rhs) const -> bool {
+    CARBON_DCHECK(dirp_ == nullptr || rhs.dirp_ == nullptr ||
+                  dirp_ == rhs.dirp_);
+    return entry_.dent_ == rhs.entry_.dent_;
+  }
+
+  auto operator*() const -> const Entry& { return entry_; }
```

Comment:
```suggestion
  auto operator*() const -> const Entry& [[clang::lifetimebound]] { return entry_; }
```

That said, `Entry` just contains a pointer, so maybe return a copy?

Also, `CARBON_DCHECK(entry_.dent_)` to catch dereffing `end()`?

Reply:
My understanding is that this is required to return a reference to conform with the iterator API requirements, which is why I have an embedded object.